### PR TITLE
planner: allow filter condition pushing down to IndexScan for prefix index

### LINKED ISSUE #21145: Allow filter condition push down to IndexScan for prefix index.
## Feature Request

Allow filter condition push down to IndexScan for prefix index.

```
mysql> create table t (a text, b text, index a_b (a(255), b(255)));
Query OK, 0 rows affected (0.01 sec)

mysql> explain select * from t where a between 'a' and 'b' and b = 'b';
+-------------------------------+---------+-----------+--------------------------+---------------------------------------------------------+
| id                            | estRows | task      | access object            | operator info                                           |
+-------------------------------+---------+-----------+--------------------------+---------------------------------------------------------+
| IndexLookUp_11                | 0.25    | root      |                          |                                                         |
| ├─IndexRangeScan_8(Build)     | 250.00  | cop[tikv] | table:t, index:a_b(a, b) | range:["a","b"], keep order:false, stats:pseudo         |
| └─Selection_10(Probe)         | 0.25    | cop[tikv] |                          | eq(test.t.b, "b"), ge(test.t.a, "a"), le(test.t.a, "b") |
|   └─TableRowIDScan_9          | 250.00  | cop[tikv] | table:t                  | keep order:false, stats:pseudo                          |
+-------------------------------+---------+-----------+--------------------------+---------------------------------------------------------+
```

The condition `b = 'b'` should be able to pushed down to `IndexScan` phase to reduce the lookup table cost.

We only need to make sure the equal condition value of `b` is less than the index column length 255.

The ideal plan should be

```
mysql> explain select * from t where a between 'a' and 'b' and b = 'b';
+--------------------------+---------+-----------+--------------------------+-------------------------------------------------+
| id                       | estRows | task      | access object            | operator info                                   |
+--------------------------+---------+-----------+--------------------------+-------------------------------------------------+
| IndexReader_7            | 0.25    | root      |                          | index:Selection_6                               |
| └─Selection_6            | 0.25    | cop[tikv] |                          | eq(test.t.b, "b")                               |
|   └─IndexRangeScan_5     | 250.00  | cop[tikv] | table:t, index:a_b(a, b) | range:["a","b"], keep order:false, stats:pseudo |
+--------------------------+---------+-----------+--------------------------+-------------------------------------------------+
```

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What problem does this PR solve?

Issue Number: close #21145 

Problem Summary:

### What is changed and how it works?

What's Changed:

The conditions like `a = "xxx"` or `b < "yyy"` can be pushed down to IndexScan for the corresponding prefix index if the length of literal string is less than the length of prefix.

How it Works:

Modify the function `splitIndexFilterConditions`, which divides conditions to IndexScan parts and TableScan parts.

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- N/A

### Release note 

- allow filter condition pushing down to IndexScan for prefix index

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -936,26 +938,90 @@ func (ds *DataSource) buildIndexMergeTableScan(prop *property.PhysicalProperty,
 	return ts, partialCost, nil
 }
 
-func indexCoveringCol(col *expression.Column, indexCols []*expression.Column, idxColLens []int) bool {
+func indexCoveringCol(col *expression.Column, constVal *expression.Constant, indexCols []*expression.Column, idxColLens []int) bool {
```
Comment (winoros) [inline planner/core/find_best_task.go:941]: maybe we could let it be called by `extractColumnsAndConstants`. So we can avoid creating new slices.

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -936,26 +938,90 @@ func (ds *DataSource) buildIndexMergeTableScan(prop *property.PhysicalProperty,
...
 	for i, indexCol := range indexCols {
-		isFullLen := idxColLens[i] == types.UnspecifiedLength || idxColLens[i] == col.RetType.Flen
-		if indexCol != nil && col.Equal(nil, indexCol) && isFullLen {
-			return true
+		if indexCol != nil && col.Equal(nil, indexCol) {
+			isFullLen := idxColLens[i] == types.UnspecifiedLength || idxColLens[i] == col.RetType.Flen
+			if isFullLen {
+				return true
+			}
+			if constVal == nil {
```
Comment (winoros) [inline planner/core/find_best_task.go:948]: The `constVal` is passed in as a parameter. We can move this check out of the for loop?

**Comment by xuyifangreeneyes** [comment]
@winoros PTAL, thanks.

**Comment by mashenjun** [comment]
Hello, I am confused with the execution plan defined in `TestConditionPushDownToIndexScanForPrefixIndex` in regards to these two following questions. 
1. The filter condition `a between 'a' and 'b'` is pushed down to `IndexRangeScan` as a range filter. Why `Selection_10` still holds the `ge` and `le` filter condition? 
2. All related columns in this query are covered by `index a_b`. Is `IndexReader` a better choice for root Task?

**Comment by xuyifangreeneyes** [comment]
> Hello, I am confused with the execution plan defined in `TestConditionPushDownToIndexScanForPrefixIndex` in regards to these two following questions.
> 
> 1. The filter condition `a between 'a' and 'b'` is pushed down to `IndexRangeScan` as a range filter. Why `Selection_10` still holds the `ge` and `le` filter condition?
> 2. All related columns in this query are covered by `index a_b`. Is `IndexReader` a better choice for root Task?

For the first question, `ge` and `le` filter condition can be deleted from `Selection_10`. I will improve it later.

For the second question, since `index a_b` is only prefix index, table scan is necessary to get full length value of `a`.

File: util/ranger/detacher.go [hunk 3]
```diff
@@ -446,6 +446,34 @@ func allEqOrIn(expr expression.Expression) bool {
 	return false
 }
 
+// accessCondShouldReserve checks whether access condition should be reserved to filter condition.
+// NOTE: access must be EQ/IN function or LogicOr of EQ/IN functions
+func accessCondShouldReserve(access expression.Expression, length int) bool {
+	f, _ := access.(*expression.ScalarFunction)
+	switch f.FuncName.L {
+	case ast.LogicOr:
+		return accessCondShouldReserve(f.GetArgs()[0], length) || accessCondShouldReserve(f.GetArgs()[0], length)
```
Comment (mashenjun) [inline util/ranger/detacher.go:455]: Hello, I didn't get the recursion logic here, why we only evaluate `f.GetArgs()[0]`?

Comment (xuyifangreeneyes) [reply util/ranger/detacher.go]: Sorry, I made a mistake. The second should be `f.GetArgs()[1]`.

**Comment by xuyifangreeneyes** [comment]
@mashenjun @winoros @lzmhhh123 PTAL

**Review by winoros — COMMENTED** [review]
lgtm

**Comment by xuyifangreeneyes** [comment]
The pr has too many conflicts so I will close it and open another one.

**Pull request closed without merging**